src/cryptonote_core/service_node_swarm.cpp

```diff
@@ -0,0 +1,270 @@
+#include "service_node_swarm.h"
+
+#undef LOKI_DEFAULT_LOG_CATEGORY
+#define LOKI_DEFAULT_LOG_CATEGORY "service_nodes"
+
+#ifdef UNIT_TEST
+  #define prod_static
+#else
+  #define prod_static static
+#endif
+
+namespace service_nodes
+{
+  static uint64_t get_new_swarm_id(std::mt19937_64 &mt, const swarm_snode_map_t &swarm_to_snodes)
+  {
+    uint64_t id_new;
+    do
+    {
+      id_new = uniform_distribution_portable(mt, UINT64_MAX);
+    } while (swarm_to_snodes.count(id_new) != 0);
+
+    return id_new;
+  }
+
+  /// The excess is calculated as the total number of snodes above MIN_SWARM_SIZE across all swarms
+  prod_static size_t calc_excess(const swarm_snode_map_t &swarm_to_snodes)
+  {
+    const size_t excess = std::accumulate(swarm_to_snodes.begin(),
+                                          swarm_to_snodes.end(),
+                                          size_t(0),
+                                          [](size_t result, const swarm_snode_map_t::value_type &pair) {
+                                            const ssize_t margin = pair.second.size() - EXCESS_BASE;
+                                            return result + std::max(margin, ssize_t(0));
+                                          });
+    LOG_PRINT_L2("Calculated excess: " << excess);
+    return excess;
+  };
+
+  /// Calculate threshold above which the excess should create a new swarm.
+  /// The threshold should be such that
+  /// 1. there is enough excess to create a new swarm of size NEW_SWARM_SIZE AND
+  /// 2. there is enough excess to leave IDEAL_SWARM_MARGIN excess in the existing swarms
+  prod_static size_t calc_threshold(const swarm_snode_map_t &swarm_to_snodes)
+  {
+    const size_t threshold = NEW_SWARM_SIZE + (swarm_to_snodes.size() * IDEAL_SWARM_MARGIN);
+    LOG_PRINT_L2("Calculated threshold: " << threshold);
+    return threshold;
+  };
+
+  prod_static void create_new_swarm_from_excess(swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt)
```

Comment:
This seems too complex. Would it be easier if you only maintained a list of swarm ids with excess nodes? (not all their snodes flattened)

Reply:
I've moved the logic int a new  `steal_from_excess_pool` function.